### ISSUE TITLE
Retain scale when resizing

### ISSUE THREAD

**Issue by medihack** [cornerstonejs/cornerstone#216]
When calling `resize` on an `enabledElement` (maybe because the layout changed or the browser window was resized) there currently seems to be no way to retain the scale of the image. When calling `resize` with `fitViewportToWindow === true` the image is resized to fit the viewport (as expected). Calling it with `fitViewportToWindow === false` then the visible part of the image may be cut off (when the resized canvas is smaller). It would be nice if there is a way to show the image in the same way it was presented before the `resize`, just larger or smaller.

**Comment by dannyrb** [comment]
It's not ideal, but you should be able to accomplish this by calculating the scale, calling `.resize()`, and restoring the scale. Is this issue about adding a way to accomplish this (it is a common use case) with a change to the API?

**Comment by swederik** [comment]
Yeah I could see this being useful.

**Comment by medihack** [comment]
@dannyrb Ok, seems to be a workaround. But somehow I would have expected that `resize(element, false)` would do this automatically as the other properties are also retained (eg. wwwc). The [OHIF viewer](http://viewer.ohif.org/) also doesn't act here reasonable. It simply resets the view (to show the whole image) when the size of a viewer or the layout changes.

**Comment by dannyrb** [comment]
@medihack, I agree that we should extend the method to allow for a preserved scale. It does feel like it should be an option (: If you have the time, I'd be happy to review/merge a PR on this!

Issues like this are interesting to me. I wonder how frequently non-developers resize there windows? 🤷‍♂️

**Comment by jdnarvaez** [comment]
@dannyrb FWIW I see rads/OR users resize viewports a lot when changing layouts and retaining one of the stacks they were displaying previously.

**Comment by medihack** [comment]
@dannyrb Yes, @jdnarvaez is right. Especially maximizing/minimizing a view and changing the layout is done very often. And one would expect to retain the correct (or better to say intuitive) scale somehow. I investigated how other DICOM viewers handle this, and it is more complex (and more diverse) than I initially would have thought (funny when I think about how often I use such a viewer at work).

Most viewers behave the following way:
- When the image was never zoomed and never panned then always fit the image to the window size (`fitToWindow(element)` in our context).
- When it was zoomed or panned then do a relative rescale.

Both points on its own are easy to implement, but the problem is to find out if the image was zoomed initially (panned isn't a problem as `viewpoint.translation` has an absolute value). My recommendation would be to let the application developer decide which strategy to use in the app, but add another resize strategy to scale relative to the window size change. That way other rescaling strategies could be also added in the future and the app developer can decide on its own which rescaling option (and when) fits best for his app (e.g. some DICOM viewers even don't rescale at all when the image was zoomed or panned before).

I could easily implement this, but the question is how we implement this in the API. In my opinion, the most beautiful solution would be to allow to provide the strategy as a parameter `resize(element, strategy)` with strategies `NO_RESCALE`, `RELATIVE_RESCALE` and `FIT_TO_WINDOW`. But this would be a breaking API change (not sure how fixed our API is). Another option would be to add a third parameter to the resize method (`resize(element, fitViewportToWindow, relativeRescale)`), but it wouldn't make much sense to set both `fitViewportToWindow` and `relativeRescale` to true.

Any opinions? I would be happy to make a PR if we can decide how to proceed.

**Comment by medihack** [comment]
@dannyrb Yes, @jdnarvaez is right. Especially maximizing/minimizing a view and changing the layout is done very often. And one would expect to retain the correct (or better to say intuitive) scale somehow. I investigated how other DICOM viewers handle this, and it is more complex (and more diverse) than I initially would have thought (funny when I think about how often I use such a viewer at work).

Most viewers behave the following way:
- When the image was never zoomed and never panned then always fit the image to the window size.
- When it was zoomed or panned then do a relative rescale.

I implemented this in the provided pull request and it seems to be quite intuitive (at least I hope so). I extracted the function `getImageSize` in its own file (from `fitToWindow`) as it is now used in `fitToWindow` and `resize` (I had to test somehow in resize if the image fits the window). I also rewrote the resize example to demonstrate the new functionality (just resize the window by dragging the little red square on the right bottom of the window).

I would be happy for suggestions and comments.
